Add Open DHC + Pedestal goal

eternalcode0 opened the pull request:
Added "Open + Ped" goal. Makes the DHC open and requires you to activate pedestal from within the DHC sanctuary.

This probably still has some problems with Pedestal Items and the DHC BK Reward settings. I haven't fully tested them with all 5 goal settings yet.

Catobat committed:
Add options for Universal Maps and Compasses

Catobat committed:
Add option for Universal Big Keys

Catobat committed:
Add option for Universal Small Keys with multiplier setting

Catobat committed:
Add setting for buyable Universal Small Keys

Catobat committed:
Make Universal Big Keys compatible with DHC BK Requirement Reward

Catobat committed:
Update presets with universal dungeon item options

Catobat committed:
Add settings for starting with small keys

Catobat committed:
Add settings for starting with big keys

Catobat committed:
Add settings for starting with heart containers and pieces

Catobat committed:
Update presets with new starting inventory settings

Catobat committed:
Refactor handling of Selection in LogicDropdown

Myth197 committed:
Add logic for buying keys

Myth197 committed:
Add Shop Key Rupee Logic as an optional setting

Catobat committed:
Implement compact settings with UI support

Catobat committed:
Add CLI support for compact settings

Catobat committed:
Add CLI support for loading cosmetics strings

deoxis9001 committed:
Update logic AccessMinishWoods 

Addition of Minish Wood Crest to the access logic, because if the Lake Crest is unattainable, Minish Wood can be accessed via the village.

deoxis9001 committed:
fix LonLon_RanchPot

deoxis9001 committed:
fix accessLonLon

Dreamie-TMC committed:
Update barlov code to make you win after opening only one chest

Myth197 committed:
fix Goron Cave cases and typos

eternalcode0 committed:
add Open DHC + Pedestal goal

Added "Open + Ped" goal. Makes the DHC open and requires you to
activate pedestal from within the DHC sanctuary.

This probably still has some problems with Pedestal Items and the DHC
BK Reward settings. I haven't fully tested them with all 5 goal
settings yet.

eternalcode0 commented on RandomizerCore/Resources/default.logic:
```diff
@@ -3760,6 +3760,15 @@ GotSwords; 	Helper; ; `GOTSWORDS`
 All_Requirements_Met;	`ALLREQUIREMENTS`;	requirementPrizeItem:Define:FirstByte, requirementPrizeSub:Define:SecondByte;	Helpers.GotSwords `GOTELEMENTS` `DHC_FIG` `DHC_AD`; `REQREWARD`
 
+!ifdef - OPENPED
+	AccessDHC; 	Helper;;
```
Not sure if this is the right way to set the DHC access or if it's the only rule that needs updating

Catobat replied:
It would be better to follow this up with an `!else` block that handles the stuff from lines 3772 to 3782. Otherwise the `AccessDHC` would get overwritten. (Or actually I am not sure what would happen, it might somehow add 2 locations with the same name, but in either case this is undesired behavior.)

eternalcode0 commented on RandomizerCore/Resources/Patches/goalHint.event:
```diff
@@ -84,13 +84,21 @@ String("You need to"); BYTE 0x0A
-#ifdef noDHC
-    String("pull the pedestal.")
+#ifdef vaatiWarp
+    String("pull the pedestal and"); BYTE 0x0A; String("defeat Vaati in sanctuary."); BYTE 0x0A;
```
Added a clear location for vaati to differentiate between the vaati in sanctuary (vaatiWarp) or regular dhc vaati.

Catobat replied:
We need to be careful with the line breaks at the end of lines 88 and 99 here, they would cause an empty line to be drawn. The texts are good!

eternalcode0 commented on RandomizerCore/Resources/Patches/newGame.event:
```diff
@@ -570,6 +571,9 @@ BYTE 0x40 //remove the messages when picking shells for the first time
 BYTE mC81 mC82 mC83 mC84 mC85 mC86 mC87 mC88 mC89 mC8A mC8B mC8C mC8D
 
+dhcGuards:
+BYTE 0x02 // move the dhc guards out of the way of the south entrance
```
Always moved the guards since they don't particularly impact anything except the vaatiWarp and openPed goals

Catobat replied:
Tiny nitpick: With Normal DHC, you can normally see the guards afer pulling the pedestal, and they only disappear after reaching the entrance room (which is the vanilla behavior). It's only visual in Rando, but I think it would be nice to retain that, which could be achieved my putting line 27 inside an `#ifdef noDHC`.

eternalcode0 commented:
I started this branch from my old Fast Vaati branch since the staging branch was giving me logic parse errors at the time that I created it. I'll solve the merge conflict when that's resolved on stanging (unless it's an issue with my environment, in which case I have no idea why)

Catobat commented on RandomizerCore/Resources/Patches/pedestal.event:
```diff
@@ -4,10 +4,19 @@
 		#undef noDHC
 	#endif
+	
+	// Remove the check for already watched the stained glass cutscene so that it's repeatable
```
This is not the best placement for this block, since the code below this can unset `openDHC` which could lead to unexpected behavior. Could you move this further down?
Additionally, have you made sure this cannot be used to win the game with the Open DHC option where you are supposed to beat Vaati? Either way, I think a define specific to ped goals should be used for the condition here instead of `openDHC`.

eternalcode0 replied:
I can move this block to a dedicated `#ifdef pedGoal` since that's defined for both openDHC and openPed.
This line only removes the check `!CheckLocalFlag(0x78)` in `sub_StateChange_Sanctuary_StainedGlass` making the vaati cutscene repeatable after watching it. It's not tied to the triggering the goal itself. The goal code starts at line 70 and is only changed to be applied to when `pedGoal` is defined (openDHC + openPed).

Catobat commented on RandomizerCore/Resources/default.logic:
```diff
@@ -3760,6 +3760,15 @@ GotSwords; 	Helper; ; `GOTSWORDS`
 
+!ifdef - OPENPED
+	AccessDHC; 	Helper;;
+	!define - DHCBK - Items.Rupee5
```
Is the intention to make Vaati inaccessible by removing the DHC Big Key?
This line only removes the DHC Big Key from the vanilla location if Big Keys are set to Vanilla. With other settings it would still get placed like normal, for that it would have to be removed from the item pool. The Requirement Reward can also still give you the Big Key (and this is used as a small "hack" for Mystery because getting the DHC Big Key in this way may for the first time reveal that the requirements of the seed are met even if the key itself is completely useless, so changing the behavior here would break this trick). And for Removed Big Keys, the DHC door should be excluded (search for `!eventdefine - dhcBigDoor`).
Unfortunately there are a lot of things to consider here.
Maybe the cleanest solution would be to replace the DHC Big Key door with a permanently closed door, then we could even leave the DHC BK in the game in all settings because it would be useless, just like we do with Removed DHC. The only other needed change then would be the Removed Big Key thing which is an easy change.

eternalcode0 replied:
Yes, the intention was to make Vaati completely unreachable. I figured removing the DHCBK was the easiest way to do it. My concern was that since DHC would still be open players would still be able to fight vaati and trigger the credits even thought that wasn't their set goal. Personally, I'd prefer to have vaati still fight-able but maybe disable the credits sequence when fighting him. I wasn't sure what the implications of that would be or even whether others would prefer that so I left the solution as it is now.
I agree that there's a lot more to consider. I'll try to play around with other possibilities to see if I can permanently close the door or remove credits on vaati.

Catobat replied:
I think making Vaati fightable without ending the game would cause a lot of confusion, unless there was a very clear indication for this in the game. So I'd prefer if he wasn't fightable in this mode (or maybe only after the game has already been beaten).

Catobat commented:
> I started this branch from my old Fast Vaati branch since the staging branch was giving me logic parse errors at the time that I created it. I'll solve the merge conflict when that's resolved on stanging (unless it's an issue with my environment, in which case I have no idea why)

This is unfortunately caused by a recent merge. It will be fixed in https://github.com/minishmaker/randomizer/pull/103 . In the meantime for testing, it should be sufficient to set Small Keys to Universal.

Catobat committed:
Fix butterfly altitude

Catobat committed:
Fix butterfly Z speed and unobtainable items with subtype 2

Catobat committed:
Fix a bad character in the logic file

Catobat committed:
Fix some key logic flexibility not getting applied with Small Keys in Own Region

Catobat committed:
Fix self-locking keychain in Deepwood Shrine with Vanilla Small Keys

Fix

Catobat committed:
Fix key logic issue in Palace of Winds with Vanilla Small Keys

Catobat committed:
Fix key logic issue in Palace of Winds with Vanilla Big Keys

Catobat committed:
Fix missing logical requirement for Melari reachability

eternalcode0 committed:
fix: remove dhc boss door in favor of blocker

eternalcode0 committed:
fix blocker pointer and add back dhc bk

a couple code review changes too:
- made the dhc guards only move for noDHC
- shifted the openDHC block

eternalcode0 commented:
Moved this to the above PR `#115` since a rebase screwed with the commit log.
